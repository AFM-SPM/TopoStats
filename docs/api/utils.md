# Utils Modules

::: topostats.utils
handler: python
options:
docstring_style: numpy
rendering:
show_signature_annotations: true
