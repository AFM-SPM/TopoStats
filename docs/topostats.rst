topostats package
=================

Subpackages
-----------

.. toctree::
   :maxdepth: 4

   topostats.logs
   topostats.tracing

Submodules
----------

.. toctree::
   :maxdepth: 4

   topostats.entry_point
   topostats.filters
   topostats.grains
   topostats.grainstats
   topostats.io
   topostats.plotting
   topostats.plottingfuncs
   topostats.processing
   topostats.run_topostats
   topostats.scars
   topostats.statistics
   topostats.theme
   topostats.thresholds
   topostats.utils
   topostats.validation

Module contents
---------------

.. automodule:: topostats
   :members:
   :undoc-members:
   :show-inheritance:
