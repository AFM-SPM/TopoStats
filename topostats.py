"""
File for tracking hard-coded variables (for now)
"""

import argparse

### List of hard-coded variables in pywgwytracing.py

# global variables
gwy_colobar_ln48 = "/module/pixmap/ztype"
gwy_linematchmethod_ln51= '/module/linematch/method'
gwy_linematchmaxdegree_ln52 = "/module/linematch/max_degree"
gwy_polylevel_coldegree_ln53 = "/module/polylevel/col_degree"

# heightediting()
datafield_grains_mark_height_ln128 = int()
gwy_polylevel_masking_ln131 = int()
gwy_linematch_masking_ln135 = int()
filterwidth_ln144 = float()

# editfile()
datafield_markheight_ln170 = int()
gwy_polylevel_masking_ln173 = int()
gwy_linematch_masking_ln177 = gwy_linematch_masking_ln135
gwy_polylevel_masking_ln181 = int()
datafield_gaussfilter_ln195 = int()

# grainfinding()
grain_gauss_ln211 = float()
datafield_mask_outliers_ln220 = float()

# remove_large_objects()
datafield_mask_outliers_ln257 = float()

# remove_small_objects()
datafield_mask_outliers_ln283 = float()

# grain_thinning()
gaussian_size_ln488 = float()

# savefiles()
savefile_type_ln535 = ["/module/pixmap/xytype"]

# saveunknownfiles()
savefile_type_ln573 = ["/module/pixmap/xytype"]

# savecroppedfiles()
savefile_type_ln645 = ["/module/pixmap/xytype"]

# __main__()
path_ln709 = str()
fileend_ln712 = []
filetype_ln713 = str()
extension_ln715 = str()
minheightscale_ln717 = float()
maxheightscale_ln718 = float()
minarea_ln722 = float()
maxdeviation_ln726 = float()
mindeviation_ln727 = float()
cropwidth_ln733 = float()
splitwidth_ln734 = float()
bins_ln736 = int()
channels_ln755 = list()


### dnatracing.py

# getNumpyArraysfromGwyddion()
sigma_ln94 = float()

# getDisorderedTrace()
sigma_ln109 = float()

# purgeObviousCrap()
maxlen_ln126 = int()

# getFittedTraces()
index_width_ln212 = int()

# getSplinedTraces()
step_size_ln376 = float()
interp_size_ln377 = float()

### tracingfuncs.py

# I didn't find any hard-coded variables so far

### statsplotting.py
### TFOplot.py
### traceplotting.py

# not checking these for now due to ongoing merging project
