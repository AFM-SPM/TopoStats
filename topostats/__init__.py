"""Topostats"""
from importlib.metadata import version

from .logs.logs import setup_logger

LOGGER = setup_logger()

release = version("topostats")
__version__ = ".".join(release.split("."[:2]))
