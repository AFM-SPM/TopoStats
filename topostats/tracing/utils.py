"""Utilities for tracing"""
from typing import Callable, Union

import numpy as np


class orderTrace:  # pylint: disable=too-few-public-methods
    """Class for ordering co-ordinates of a skeleton"""

    def __init__(self, coordinates: Union[np.ndarray, list]):
        """Initialise the class."""
        self.coordinates = coordinates

    def order(self, shape: str) -> np.ndarray:
        """Order a trace of the given shape.

        Parameters
        ----------
        shape: str
            The shape of the molecule, should be either 'linear' or 'circle'.capitalize()

        Returns
        -------
        Callable"""
        return self._order(shape)

    def _order(self, shape: str) -> Callable:
        """Creator for the method of ordering to use.

        Parameters
        ----------
        shape : str
            The shape of the molecule, should be either 'linear' or 'circle'.

        Returns
        -------
        Callable
            Returns the appropriate method for ordering the coordinates.

        """
        if shape == "linear":
            return self._linear()
        if shape == "circle":
            return self._circle()
        raise ValueError(shape)

    def _linear(self) -> np.ndarray:
        """Order linear skeleton."""

    def _circle(self) -> np.ndarray:
        """Order circular skeleton."""

    #    def _dilate(self):

    def _find_end_point(self):
        """Find end-points of linear grains."""


class Neighbours:  # pylint: disable=too-few-public-methods
    """Class for summarising features of neighbouring pixels."""

    def __init__(self):
        """Initialise the class."""
