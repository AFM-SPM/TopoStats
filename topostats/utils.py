"""Utilities"""
from argparse import Namespace
import logging
from pathlib import Path
from typing import Union, Dict
from collections import defaultdict

import numpy as np
import pandas as pd

from topostats.thresholds import threshold
from topostats.logs.logs import LOGGER_NAME

LOGGER = logging.getLogger(LOGGER_NAME)


ALL_STATISTICS_COLUMNS = (
    "image",
    "basename",
    "molecule_number",
    "area",
    "area_cartesian_bbox",
    "aspect_ratio",
    "bending_angle",
    "centre_x",
    "centre_y",
    "circular",
    "contour_lengths",
    "end_to_end_distance",
    "height_max",
    "height_mean",
    "height_median",
    "height_min",
    "max_feret",
    "min_feret",
    "radius_max",
    "radius_mean",
    "radius_median",
    "radius_min",
    "smallest_bounding_area",
    "smallest_bounding_length",
    "smallest_bounding_width",
    "threshold",
    "volume",
)


def convert_path(path: Union[str, Path]) -> Path:
    """Ensure path is Path object.

    Parameters
    ----------
    path: Union[str, Path]
        Path to be converted.

    Returns
    -------
    Path
        pathlib Path
    """
    return Path().cwd() if path == "./" else Path(path).expanduser()


def update_config(config: dict, args: Union[dict, Namespace]) -> Dict:
    """Update the configuration with any arguments

    Parameters
    ----------
    config: dict
        Dictionary of configuration (typically read from YAML file specified with '-c/--config <filename>')
    args: Namespace
        Command line arguments
    Returns
    -------
    Dict
        Dictionary updated with command arguments.
    """
    args = vars(args) if isinstance(args, Namespace) else args

    config_keys = config.keys()
    for arg_key, arg_value in args.items():
        if isinstance(arg_value, dict):
            update_config(config, arg_value)
        else:
            if arg_key in config_keys and arg_value is not None:
                original_value = config[arg_key]
                config[arg_key] = arg_value
                LOGGER.info(f"Updated config config[{arg_key}] : {original_value} > {arg_value} ")
    if "base_dir" in config.keys():
        config["base_dir"] = convert_path(config["base_dir"])
    if "output_dir" in config.keys():
        config["output_dir"] = convert_path(config["output_dir"])
    return config


def update_plotting_config(plotting_config: dict) -> dict:
    """Update the plotting config for each of the plots in plot_dict to ensure that each
    entry has all the plotting configuration values that are needed."""

    main_config = plotting_config.copy()
    for opt in ["plot_dict", "run"]:
        main_config.pop(opt)
    for image, options in plotting_config["plot_dict"].items():
        plotting_config["plot_dict"][image] = {**options, **main_config}
        if image not in ["z_threshed", "mask_overlay"]:
            plotting_config["plot_dict"][image].pop("zrange")

    return plotting_config


def _get_mask(image: np.ndarray, thresh: float, threshold_direction: str, img_name: str = None) -> np.ndarray:
    """Calculate a mask for pixels that exceed the threshold

    Parameters
    ----------
    image: np.array
        Numpy array representing image.
    threshold: float
        A float representing the threshold
    threshold_direction: str
        A string representing the direction that should be thresholded. ("above", "below")
    img_name: str
        Name of image being processed

    Returns
    -------
    np.array
        Numpy array of image with objects coloured.
    """
    if threshold_direction == "above":
        LOGGER.info(f"[{img_name}] : Masking (above) Threshold: {thresh}")
        return image > thresh
    LOGGER.info(f"[{img_name}] : Masking (below) Threshold: {thresh}")
    return image < thresh
    # LOGGER.fatal(f"[{img_name}] : Threshold direction invalid: {threshold_direction}")


def get_mask(image: np.ndarray, thresholds: dict, img_name: str = None) -> np.ndarray:
    """Mask data that should not be included in flattening.

    Parameters
    ----------
    image: np.ndarray
        2D Numpy array of the image to have a mask derived for.

    thresholds: dict
        Dictionary of thresholds, at a bare minimum must have key 'below' with an associated value, second key is
        to have an 'above' threshold.
    img_name: str
        Image name that is being masked.

    Returns
    -------
    np.ndarray
        2D Numpy boolean array of points to mask.
    """
    # Both thresholds are applicable
    if "below" in thresholds and "above" in thresholds:
        mask_above = _get_mask(image, thresh=thresholds["above"], threshold_direction="above", img_name=img_name)
        mask_below = _get_mask(image, thresh=thresholds["below"], threshold_direction="below", img_name=img_name)
        # Masks are combined to remove both the extreme high and extreme low data points.
        return mask_above + mask_below
    # Only below threshold is applicable
    if "below" in thresholds:
        return _get_mask(image, thresh=thresholds["below"], threshold_direction="below", img_name=img_name)
    # Only above threshold is applicable
    return _get_mask(image, thresh=thresholds["above"], threshold_direction="above", img_name=img_name)


# pylint: disable=unused-argument
def get_thresholds(
    image: np.ndarray,
    threshold_method: str,
    otsu_threshold_multiplier: float = None,
    threshold_std_dev: dict = None,
    absolute: dict = None,
    **kwargs,
) -> Dict:
    """Obtain thresholds for masking data points.

    Parameters
    ----------
    image : np.ndarray
        2D Numpy array of image to be masked
    threshold_method : str
        Method for thresholding, 'otsu', 'std_dev' or 'absolute' are valid options.
    threshold_std_dev : dict
        Dict of above and below thresholds for the standard deviation method.
    absolute : tuple
        Dict of below and above thresholds.
    **kwargs:

    Returns
    -------
    Dict
        Dictionary of thresholds, contains keys 'below' and optionally 'above'.
    """
    thresholds = defaultdict()
    if threshold_method == "otsu":
        thresholds["above"] = threshold(image, method="otsu", otsu_threshold_multiplier=otsu_threshold_multiplier)
    elif threshold_method == "std_dev":
        try:
            if threshold_std_dev["below"] is not None:
                thresholds["below"] = threshold(image, method="mean") - threshold_std_dev["below"] * np.nanstd(image)
            if threshold_std_dev["above"] is not None:
                thresholds["above"] = threshold(image, method="mean") + threshold_std_dev["above"] * np.nanstd(image)
        except TypeError as typeerror:
            raise typeerror
    elif threshold_method == "absolute":
        if absolute["below"] is not None:
            thresholds["below"] = absolute["below"]
        if absolute["above"] is not None:
            thresholds["above"] = absolute["above"]
    else:
        if not isinstance(threshold_method, str):
            raise TypeError(
                f"threshold_method ({threshold_method}) should be a string. Valid values : 'otsu' 'std_dev' 'absolute'"
            )
        if threshold_method not in ["otsu", "std_dev", "absolute"]:
            raise ValueError(
                f"threshold_method ({threshold_method}) is invalid. Valid values : 'otsu' 'std_dev' 'absolute'"
            )
    return thresholds


def create_empty_dataframe(columns: set = ALL_STATISTICS_COLUMNS, index: tuple = "molecule_number") -> pd.DataFrame:
    """Create an empty data frame for returning when no results are found.

    Parameters
    ----------
    columns: list
        Columns of the empty dataframe.

    Returns
    -------
    pd.DataFrame
        Empty Pandas DataFrame.
    """
    empty_df = pd.DataFrame(columns=columns)
    empty_df = empty_df.set_index(index)
    return empty_df
