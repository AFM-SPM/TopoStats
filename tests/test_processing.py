"""Test end-to-end running of topostats."""
from pathlib import Path

# pylint: disable=deprecated-module
import imghdr
import numpy as np
import pytest

from topostats.io import LoadScans
from topostats.processing import check_run_steps, process_scan
from topostats.utils import update_plotting_config

BASE_DIR = Path.cwd()


# Can't see a way of paramterising with pytest-regtest as it writes to a file based on the file/function
# so instead we run three regression tests.
def test_process_scan_below(regtest, tmp_path, process_scan_config: dict, load_scan_data: LoadScans) -> None:
    """Regression test for checking the process_scan functions correctly"""
    process_scan_config["grains"]["direction"] = "below"
    img_dic = load_scan_data.img_dict
    _, results = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )
    # Remove the basename column as this differs on CI
    results.drop(["basename"], axis=1, inplace=True)
    print(results.to_string(), file=regtest)  # noqa: T201


def test_process_scan_above(regtest, tmp_path, process_scan_config: dict, load_scan_data: LoadScans) -> None:
    """Regression test for checking the process_scan functions correctly"""
    img_dic = load_scan_data.img_dict
    _, results = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )
    # Remove the Basename column as this differs on CI
    results.drop(["basename"], axis=1, inplace=True)
    print(results.to_string(), file=regtest)  # noqa: T201


def test_process_scan_both(regtest, tmp_path, process_scan_config: dict, load_scan_data: LoadScans) -> None:
    """Regression test for checking the process_scan functions correctly"""
    process_scan_config["grains"]["direction"] = "both"
    img_dic = load_scan_data.img_dict
    _, results = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )
    # Remove the Basename column as this differs on CI
    results.drop(["basename"], axis=1, inplace=True)
    print(results.to_string(), file=regtest)  # noqa: T201


@pytest.mark.parametrize(
    "image_set, expected",
    [
        ("core", False),
        ("all", True),
    ],
)
def test_save_cropped_grains(
    tmp_path: Path, process_scan_config: dict, load_scan_data: LoadScans, image_set, expected
) -> None:
    """Tests if cropped grains are saved only when image set is 'all' rather than 'core'."""
    process_scan_config["plotting"]["image_set"] = image_set
    process_scan_config["plotting"] = update_plotting_config(process_scan_config["plotting"])

    img_dic = load_scan_data.img_dict
    _, _ = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )

    assert (
        Path.exists(tmp_path / "tests/resources/processed/minicircle/grains/above" / "minicircle_grain_image_0.png")
        == expected
    )
    assert (
        Path.exists(tmp_path / "tests/resources/processed/minicircle/grains/above" / "minicircle_grain_mask_0.png")
        == expected
    )
    assert (
        Path.exists(
            tmp_path / "tests/resources/processed/minicircle/grains/above" / "minicircle_grain_mask_image_0.png"
        )
        == expected
    )


@pytest.mark.parametrize("extension", [("png"), ("tiff")])
def test_save_format(process_scan_config: dict, load_scan_data: LoadScans, tmp_path: Path, extension: str):
    """Tests if save format applied to cropped images"""
    process_scan_config["plotting"]["image_set"] = "all"
    process_scan_config["plotting"]["save_format"] = extension
    process_scan_config["plotting"] = update_plotting_config(process_scan_config["plotting"])

    img_dic = load_scan_data.img_dict
    _, _ = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )

    assert (
        imghdr.what(
            tmp_path / "tests/resources/processed/minicircle/grains/above" / f"minicircle_grain_image_0.{extension}"
        )
        == extension
    )


@pytest.mark.parametrize(
    "filter_run, grains_run, grainstats_run, dnatracing_run, log_msg",
    [
        (
            False,
            False,
            False,
            True,
            "DNA tracing enabled but Grainstats disabled. Please check your configuration file.",
        ),
        (
            False,
            False,
            True,
            True,
            "DNA tracing enabled but Grains disabled. Please check your configuration file.",
        ),
        (
            False,
            True,
            True,
            True,
            "DNA tracing enabled but Filters disabled. Please check your configuration file.",
        ),
        (
            False,
            False,
            True,
            False,
            "Grainstats enabled but Grains disabled. Please check your configuration file.",
        ),
        (
            False,
            True,
            True,
            False,
            "Grainstats enabled but Filters disabled. Please check your configuration file.",
        ),
        (
            False,
            True,
            False,
            False,
            "Grains enabled but Filters disabled. Please check your configuration file.",
        ),
        (
            True,
            False,
            False,
            False,
            "Configuration run options are consistent, processing can proceed.",
        ),
        (
            True,
            True,
            False,
            False,
            "Configuration run options are consistent, processing can proceed.",
        ),
        (
            True,
            True,
            True,
            False,
            "Configuration run options are consistent, processing can proceed.",
        ),
        (
            True,
            True,
            True,
            True,
            "Configuration run options are consistent, processing can proceed.",
        ),
    ],
)
def test_check_run_steps(
    filter_run: bool,
    grains_run: bool,
    grainstats_run: bool,
    dnatracing_run: bool,
    log_msg: str,
    caplog,
) -> None:
    """Test the logic which checks whether enabled processing options are consistent."""
    check_run_steps(filter_run, grains_run, grainstats_run, dnatracing_run)
    assert log_msg in caplog.text


# noqa: disable=too-many-arguments
# pylint: disable=too-many-arguments
@pytest.mark.parametrize(
    "filter_run, grains_run, grainstats_run, dnatracing_run, log_msg1, log_msg2",
    [
        (
            False,
            False,
            False,
            False,
            "You have not included running the initial filter stage.",
            "Please check your configuration file.",
        ),
        (
            True,
            False,
            False,
            False,
            "Detection of grains disabled, returning empty data frame.",
            "13-gaussian_filtered",
        ),
        (
            True,
            True,
            False,
            False,
            "Calculation of grainstats disabled, returning empty data frame.",
            "22-labelled_image_bboxes",
        ),
        (
            True,
            True,
            True,
            False,
            "Processing grain",
            "Calculation of DNA Tracing disabled, returning grainstats data frame.",
        ),
        (
            True,
            True,
            True,
            True,
            "There are 15 circular and 6 linear DNA molecules found in the image",
            "Combining above grain statistics and dnatracing statistics",
        ),
    ],
)
def test_process_stages(
    process_scan_config: dict,
    load_scan_data: LoadScans,
    tmp_path: Path,
    filter_run: bool,
    grains_run: bool,
    grainstats_run: bool,
    dnatracing_run: bool,
    log_msg1: str,
    log_msg2: str,
    caplog,
) -> None:
    """Regression test for checking the process_scan functions correctly when specific sections are disabled.

    Currently there is no test for having later stages (e.g. DNA Tracing or Grainstats) enabled when Filters and/or
    Grainstats are disabled. Whislt possible it is expected that users understand the need to run earlier stages before
    later staged can run and do not disable earlier stages."""
    img_dic = load_scan_data.img_dict
    process_scan_config["filter"]["run"] = filter_run
    process_scan_config["grains"]["run"] = grains_run
    process_scan_config["grainstats"]["run"] = grainstats_run
    process_scan_config["dnatracing"]["run"] = dnatracing_run
    _, _ = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )

    assert log_msg1 in caplog.text
    assert log_msg2 in caplog.text


def test_process_scan_no_grains(process_scan_config: dict, load_scan_data: LoadScans, tmp_path: Path, caplog) -> None:
    """Test handling no grains found during grains.find_grains()."""
    img_dic = load_scan_data.img_dict
    process_scan_config["grains"]["threshold_std_dev"]["above"] = 1000
    process_scan_config["filter"]["remove_scars"]["run"] = False
    _, _ = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )
    assert "Grains found for direction above : 0" in caplog.text
    assert "There are 0 circular and 0 linear DNA molecules found in the image" in caplog.text
    assert "No grains exist for the above direction. Skipping grainstats and DNAtracing." in caplog.text


def test_process_scan_align_grainstats_dnatracing(
    process_scan_config: dict, load_scan_data: LoadScans, tmp_path: Path
) -> None:
    """Test that molecule numbers from dnatracing align with those from grainstats when some grains are removed from
    tracing because they are too small.

    By setting processing parameters as below two molecules are pruged for being too small after skeletonisation and so
    do not have DNA tracing statistics (but they do have Grain Statistics)."""
    img_dic = load_scan_data.img_dict
    process_scan_config["filter"]["remove_scars"]["run"] = False
    process_scan_config["grains"]["absolute_area_threshold"]["above"] = [150, 3000]
    process_scan_config["dnatracing"]["min_skeleton_size"] = 10
    _, results = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )
    tracing_to_check = ["contour_lengths", "circular", "end_to_end_distance"]
    assert results.shape == (24, 25)
    assert np.isnan(results.loc[8, "contour_lengths"])
    assert np.isnan(sum(results.loc[8, tracing_to_check]))
    assert np.isnan(results.loc[22, "contour_lengths"])
    assert np.isnan(sum(results.loc[22, tracing_to_check]))
