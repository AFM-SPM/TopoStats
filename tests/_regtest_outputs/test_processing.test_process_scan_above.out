                     centre_x      centre_y    radius_min    radius_max   radius_mean  radius_median    height_min    height_max  height_median   height_mean        volume          area  area_cartesian_bbox  smallest_bounding_width  smallest_bounding_length  smallest_bounding_area  aspect_ratio threshold     max_feret     min_feret       image  contour_lengths  circular  end_to_end_distance
molecule_number                                                                                                                                                                                                                                                                                                                                                                                          
0                1.872403e-08  9.654910e-08  5.266374e-09  2.102504e-08  1.535118e-08   1.601571e-08  7.821915e-10  3.932379e-09   2.101130e-09  2.005276e-09  1.282137e-24  6.393819e-16         1.137465e-15             2.405749e-08              4.124140e-08            9.921645e-16      1.714286     above  4.138436e-08  2.404835e-08  minicircle     4.671362e-08      True         0.000000e+00
1                4.002647e-07  9.315342e-08  1.161835e-09  2.193371e-08  1.378055e-08   1.472411e-08  7.825457e-10  3.194750e-09   2.169921e-09  2.023310e-09  1.278361e-24  6.318167e-16         1.186029e-15             2.792782e-08              3.962207e-08            1.106558e-15      1.418731     above  3.967123e-08  2.792782e-08  minicircle     1.152444e-07     False         2.147067e-08
2                2.178216e-07  1.731429e-07  7.874120e-09  2.883227e-08  1.778069e-08   1.633758e-08  7.819369e-10  3.135523e-09   2.015185e-09  1.900232e-09  1.184832e-24  6.235194e-16         1.592842e-15             2.416820e-08              5.458082e-08            1.319120e-15      2.258373     above  5.469842e-08  2.416820e-08  minicircle     9.024501e-08      True         0.000000e+00
3                3.534386e-07  2.079919e-07  3.994787e-09  2.569119e-08  1.556160e-08   1.628550e-08  7.818716e-10  3.112567e-09   2.154862e-09  2.025576e-09  1.233327e-24  6.088770e-16         1.619442e-15             1.793469e-08              4.959678e-08            8.895027e-16      2.765411     above  4.968108e-08  1.793469e-08  minicircle     8.351526e-08     False         2.359349e-08
4                3.013918e-07  2.116190e-07  9.783923e-09  2.529492e-08  1.787818e-08   1.842165e-08  7.820739e-10  3.007900e-09   1.920090e-09  1.863138e-09  1.346757e-24  7.228432e-16         1.739509e-15             3.332044e-08              4.395027e-08            1.464442e-15      1.319018     above  4.418497e-08  3.332044e-08  minicircle     9.591161e-08      True         0.000000e+00
5                4.325016e-07  2.158008e-07  2.404072e-09  2.355884e-08  1.505586e-08   1.590245e-08  7.824373e-10  3.242017e-09   2.197200e-09  2.047442e-09  1.261630e-24  6.161982e-16         1.259241e-15             2.065823e-08              4.594861e-08            9.492169e-16      2.224228     above  4.620710e-08  2.065823e-08  minicircle     9.292059e-08     False         2.253542e-08
6                1.944673e-07  2.311843e-07  8.956326e-09  2.396012e-08  1.797012e-08   1.828566e-08  7.829192e-10  2.930014e-09   2.070243e-09  1.944491e-09  1.341501e-24  6.898980e-16         1.559409e-15             2.974231e-08              4.527651e-08            1.346628e-15      1.522293     above  4.625725e-08  2.974231e-08  minicircle     9.745305e-08      True         0.000000e+00
7                3.897351e-07  2.743875e-07  1.613377e-09  2.554789e-08  1.652770e-08   1.806023e-08  7.819917e-10  3.053690e-09   2.228219e-09  2.052587e-09  1.445129e-24  7.040522e-16         1.770258e-15             2.241199e-08              4.910628e-08            1.100570e-15      2.191071     above  4.922213e-08  2.241199e-08  minicircle     4.243137e-08      True         0.000000e+00
8                4.304040e-07  2.959393e-07  2.342765e-09  2.229247e-08  1.526636e-08   1.660954e-08  7.823310e-10  4.416205e-09   2.218856e-09  2.099831e-09  1.390251e-24  6.620775e-16         1.420794e-15             2.691722e-08              4.172411e-08            1.123097e-15      1.550089     above  4.354237e-08  2.606386e-08  minicircle     5.672940e-08      True         0.000000e+00
9                1.227373e-07  3.088215e-07  8.467644e-09  2.745279e-08  1.765053e-08   1.727278e-08  7.823695e-10  3.059652e-09   2.264351e-09  2.068486e-09  1.431083e-24  6.918503e-16         1.706320e-15             2.555373e-08              5.220711e-08            1.334087e-15      2.043033     above  5.246906e-08  2.555373e-08  minicircle     7.586297e-08      True         0.000000e+00
10               3.889252e-08  3.823969e-07  1.694688e-09  2.774991e-08  1.663512e-08   1.696498e-08  7.824603e-10  2.966891e-09   2.212553e-09  2.038431e-09  1.360046e-24  6.672023e-16         1.234837e-15             1.952555e-08              5.456852e-08            1.065480e-15      2.794724     above  5.457559e-08  1.952555e-08  minicircle     5.187387e-08      True         0.000000e+00
11               3.897261e-07  3.775629e-07  9.035783e-09  2.695127e-08  1.742859e-08   1.679317e-08  7.824575e-10  3.050728e-09   2.198785e-09  2.024155e-09  1.346569e-24  6.652500e-16         1.464233e-15             2.837678e-08              4.850977e-08            1.376551e-15      1.709488     above  4.902839e-08  2.837678e-08  minicircle     8.282483e-08      True         0.000000e+00
12               2.612096e-07  3.748066e-07  5.293707e-09  2.546113e-08  1.740346e-08   1.836583e-08  7.820596e-10  2.941901e-09   2.101283e-09  1.955868e-09  1.361759e-24  6.962430e-16         1.452520e-15             2.556090e-08              5.001954e-08            1.278544e-15      1.956878     above  5.018690e-08  2.556090e-08  minicircle     1.002724e-07      True         0.000000e+00
13               3.046740e-07  3.910695e-07  1.089194e-08  2.666028e-08  1.829150e-08   1.850468e-08  7.831113e-10  3.089554e-09   2.050870e-09  1.929329e-09  1.293844e-24  6.706189e-16         1.665565e-15             3.552953e-08              4.206139e-08            1.494421e-15      1.183843     above  4.904581e-08  3.313363e-08  minicircle     9.867118e-08      True         0.000000e+00
14               4.555221e-07  4.280578e-07  4.268437e-09  2.681750e-08  1.659015e-08   1.655964e-08  7.844241e-10  3.819751e-09   2.213004e-09  2.065125e-09  1.357194e-24  6.571967e-16         1.319762e-15             2.194042e-08              5.174054e-08            1.135209e-15      2.358230     above  5.174076e-08  2.194042e-08  minicircle     1.068649e-07     False         1.630958e-08
15               2.204550e-07  4.181537e-07  3.939995e-09  2.749079e-08  1.683256e-08   1.618414e-08  7.823262e-10  3.206657e-09   2.127363e-09  1.983822e-09  1.379770e-24  6.955108e-16         1.530124e-15             2.535473e-08              5.126658e-08            1.299850e-15      2.021973     above  5.148307e-08  2.535473e-08  minicircle     9.138358e-08      True         0.000000e+00
16               3.761571e-07  4.281550e-07  7.920343e-09  2.490765e-08  1.874191e-08   1.934915e-08  7.832919e-10  3.105480e-09   2.140375e-09  1.992034e-09  1.428747e-24  7.172303e-16         1.530124e-15             3.161664e-08              4.678904e-08            1.479313e-15      1.479886     above  4.805051e-08  3.161664e-08  minicircle     1.064535e-07      True         0.000000e+00
17               4.608434e-07  4.586829e-07  8.004359e-09  2.780169e-08  1.726742e-08   1.679523e-08  7.825776e-10  3.150258e-09   2.184948e-09  2.018464e-09  1.288599e-24  6.384057e-16         1.616514e-15             2.133895e-08              5.270820e-08            1.124738e-15      2.470046     above  5.325388e-08  2.133895e-08  minicircle     6.992770e-08      True         0.000000e+00
18               4.198202e-07  4.563309e-07  1.015119e-08  2.481940e-08  1.821840e-08   1.883696e-08  7.836888e-10  3.073389e-09   2.164871e-09  2.006525e-09  1.387725e-24  6.916062e-16         1.603336e-15             3.682726e-08              4.142321e-08            1.525503e-15      1.124797     above  4.596617e-08  3.355597e-08  minicircle     1.017678e-07      True         0.000000e+00
19               9.101179e-08  4.690084e-07  4.483187e-09  2.542198e-08  1.635227e-08   1.694269e-08  7.830648e-10  3.024658e-09   2.183333e-09  2.027181e-09  1.369855e-24  6.757437e-16         1.600163e-15             2.099968e-08              4.888798e-08            1.026632e-15      2.328035     above  4.891627e-08  2.099968e-08  minicircle     9.998428e-08     False         1.578492e-08
20               2.032140e-08  5.007542e-07  1.558019e-09  2.537412e-08  1.581556e-08   1.625252e-08  7.860746e-10  3.083383e-09   2.222899e-09  2.055487e-09  1.289160e-24  6.271800e-16         1.391022e-15             2.035424e-08              5.020435e-08            1.021871e-15      2.466530     above  5.020878e-08  2.035424e-08  minicircle     9.787399e-08     False         2.096459e-08
